README.md
# react-img-atv

![Last version](https://img.shields.io/github/tag/Kikobeats/react-img-atv.svg?style=flat-square)
[![Build Status](http://img.shields.io/travis/Kikobeats/react-img-atv/master.svg?style=flat-square)](https://travis-ci.org/Kikobeats/react-img-atv)
[![Dependency status](http://img.shields.io/david/Kikobeats/react-img-atv.svg?style=flat-square)](https://david-dm.org/Kikobeats/react-img-atv)
[![Dev Dependencies Status](http://img.shields.io/david/dev/Kikobeats/react-img-atv.svg?style=flat-square)](https://david-dm.org/Kikobeats/react-img-atv#info=devDependencies)
[![Donate](https://img.shields.io/badge/donate-paypal-blue.svg?style=flat-square)](https://paypal.me/kikobeats)
> Apple TV 3D parallax effect as component

## Install

`npm install --save react-img-atv`

## Usage

```jsx
import AtvImg from 'react-img-atv';

<AtvImg
  layers={[
    'http://kloc.pm/images/back.png',
    'http://kloc.pm/images/front.png',
  ]}
  staticFallback="http://kloc.pm/images/kloc-icon-flattened.jpg"
  isStatic={false}
  className={'thisIsOptional'}
  style={{ width: 320, height: 190 }}
/>
```

## API

### props

``` jsx
static propTypes = {
  layers: PropTypes.arrayOf(PropTypes.string).isRequired,
  isStatic: PropTypes.bool,
  staticFallback: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.object,
};
```

#### data

- `layers`: Required. An array of image URLs. The images will be stacked on top of each other and the _last_ element will be at the top.
- `isStatic`: Optional. A boolean value. When it evaluates to `true`, it disables the parallax effect, and shows the flattened image (`staticFallback`) instead.
- `staticFallback`: Optional. A flattened image that contains all the layers.

#### styling

Use the following options to set up the width/height of the entire component

- `className`: Optional.
- `style`: Optional.

## License

**react-img-atv** © [Kiko Beats](https://kikobeats.com), Released under the [MIT](https://github.com/Kikobeats/react-img-atv/blob/master/LICENSE.md) License.<br>
Authored and maintained by Kiko Beats with help from [contributors](https://github.com/Kikobeats/react-img-atv/contributors).

> [kikobeats.com](https://kikobeats.com) · GitHub [Kiko Beats](https://github.com/kikobeats) · Twitter [@kikobeats](https://twitter.com/kikobeats)


CHANGELOG.md
<a name="1.0.3"></a>
## 1.0.3 (2018-04-26)

* Add better browser targets ([d193887](https://github.com/Kikobeats/react-img-atv/commit/d193887))
* Update package.json ([fab37fe](https://github.com/Kikobeats/react-img-atv/commit/fab37fe))
* Update README.md ([c3987ec](https://github.com/Kikobeats/react-img-atv/commit/c3987ec))
* fix(package): update prop-types to version 15.6.1 ([ea54cda](https://github.com/Kikobeats/react-img-atv/commit/ea54cda))
* docs(readme): add Greenkeeper badge ([ea102fb](https://github.com/Kikobeats/react-img-atv/commit/ea102fb))



<a name="1.0.2"></a>
## 1.0.2 (2018-01-01)

* Move from engine to peerDependency ([2bc6063](https://github.com/Kikobeats/react-img-atv/commit/2bc6063))



<a name="1.0.1"></a>
## 1.0.1 (2018-01-01)

* Declare react & react-dom as dev/peer dependencies ([732cb7d](https://github.com/Kikobeats/react-img-atv/commit/732cb7d))
* Update package.json ([27a9f0e](https://github.com/Kikobeats/react-img-atv/commit/27a9f0e))
* Update README.md ([c292de7](https://github.com/Kikobeats/react-img-atv/commit/c292de7))
* Update README.md ([864199c](https://github.com/Kikobeats/react-img-atv/commit/864199c))
* chore(package): update babel-eslint to version 8.1.0 ([48880ad](https://github.com/Kikobeats/react-img-atv/commit/48880ad))
* fix(package): update react to version 16.1.0 ([4df6835](https://github.com/Kikobeats/react-img-atv/commit/4df6835))
* fix(package): update react to version 16.2.0 ([7f87e1c](https://github.com/Kikobeats/react-img-atv/commit/7f87e1c))
* fix(package): update react-dom to version 16.1.0 ([5b75e8d](https://github.com/Kikobeats/react-img-atv/commit/5b75e8d))
* fix(package): update react-dom to version 16.2.0 ([02c0c1e](https://github.com/Kikobeats/react-img-atv/commit/02c0c1e))
* docs(readme): add Greenkeeper badge ([ef88222](https://github.com/Kikobeats/react-img-atv/commit/ef88222))
* docs(readme): add Greenkeeper badge ([27318c9](https://github.com/Kikobeats/react-img-atv/commit/27318c9))



<a name="1.0.0"></a>
# 1.0.0 (2017-10-12)

* First commit ([8004ce9](https://github.com/Kikobeats/react-img-atv/commit/8004ce9))





src/styles.js
export default {
  root: {
    borderRadius: 5,
    transformStyle: 'preserve-3d',
    WebkitTapHighlightColor: 'rgba(#000, 0)'
  },

  staticFallback: {
    maxWidth: '100%',
    maxHeight: '100%',
    borderRadius: 5,
    boxShadow: '0 2px 8px rgba(14, 21, 47, 0.25)'
  },

  container: {
    position: 'relative',
    width: '100%',
    height: '100%',
    borderRadius: 5,
    transition: 'all 0.2s ease-out'
  },

  shadowOnHover: {
    boxShadow: '0 45px 100px rgba(14, 21, 47, 0.4), 0 16px 40px rgba(14, 21, 47, 0.4)'
  },

  layers: {
    position: 'relative',
    width: '100%',
    height: '100%',
    borderRadius: 5,
    overflow: 'hidden',
    transformStyle: 'preserve-3d'
  },

  renderedLayer: {
    position: 'absolute',
    width: '104%',
    height: '104%',
    top: '-2%',
    left: '-2%',
    backgroundRepeat: 'no-repeat',
    backgroundPosition: 'center',
    backgroundColor: 'transparent',
    backgroundSize: 'cover',
    transition: 'all 0.1s ease-out'
  },

  shadow: {
    position: 'absolute',
    top: '5%',
    left: '5%',
    width: '90%',
    height: '90%',
    transition: 'all 0.2s ease-out',
    boxShadow: '0 8px 30px rgba(14, 21, 47, 0.6)'
  },

  shine: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    borderRadius: 5,
    background: 'linear-gradient(135deg, rgba(255, 255, 255, .25) 0%, rgba(255, 255, 255, 0) 60%)'
  }
}


src/index.js
import React, {Component} from 'react'
import PropTypes from 'prop-types'

import styles from './styles'

export default class AtvImg extends Component {
  static propTypes = {
    layers: PropTypes.arrayOf(PropTypes.string).isRequired,
    isStatic: PropTypes.bool,
    staticFallback: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.object
  }

  state = {
    rootElemWidth: 0,
    rootElemHeight: 0,
    isOnHover: false,
    container: {},
    shine: {},
    layers: []
  }

  componentDidMount = () => {
    if (!this.props.isStatic) {
      this.setState({
        // eslint-disable-line react/no-did-mount-set-state
        // this is a legit use case. we must trigger a re-render. don't worry.
        rootElemWidth:
          this.node.clientWidth ||
          this.node.offsetWidth ||
          this.node.scrollWidth,
        rootElemHeight:
          this.node.clientHeight ||
          this.node.offsetHeight ||
          this.node.scrollHeight
      })
    }
  }

  handleMove = ({pageX, pageY}) => {
    const layerCount = this.props.layers.length // the number of layers

    const {rootElemWidth, rootElemHeight} = this.state

    const bodyScrollTop =
      document.body.scrollTop ||
      document.getElementsByTagName('html')[0].scrollTop
    const bodyScrollLeft = document.body.scrollLeft
    const offsets = this.node.getBoundingClientRect()
    const wMultiple = 320 / rootElemWidth
    const offsetX =
      0.52 - (pageX - offsets.left - bodyScrollLeft) / rootElemWidth // cursor position X
    const offsetY =
      0.52 - (pageY - offsets.top - bodyScrollTop) / rootElemHeight // cursor position Y
    const dy = pageY - offsets.top - bodyScrollTop - rootElemHeight / 2 // center Y of container
    const dx = pageX - offsets.left - bodyScrollLeft - rootElemWidth / 2 // center X of container
    const yRotate = (offsetX - dx) * (0.07 * wMultiple) // rotation for container Y
    const xRotate = (dy - offsetY) * (0.1 * wMultiple) // rotation for container X

    const arad = Math.atan2(dy, dx) // angle between cursor and center of container in RAD

    const rawAngle = arad * 180 / Math.PI - 90 // convert rad to degrees
    const angle = rawAngle < 0 ? rawAngle + 360 : rawAngle

    this.setState({
      container: {
        transform:
          `rotateX(${xRotate}deg) rotateY(${yRotate}deg)` +
          (this.state.isOnHover ? ' scale3d(1.07,1.07,1.07)' : '')
      },
      shine: {
        background: `linear-gradient(${angle}deg, rgba(255, 255, 255, ${(pageY -
          offsets.top -
          bodyScrollTop) /
          rootElemHeight *
          0.4}) 0%, rgba(255, 255, 255, 0) 80%)`,
        transform: `translateX(${offsetX * layerCount -
          0.1}px) translateY(${offsetY * layerCount - 0.1}px)`
      },
      layers: this.props.layers.map((_, idx) => ({
        transform: `translateX(${offsetX *
          (layerCount - idx) *
          (idx * 2.5 / wMultiple)}px) translateY(${offsetY *
          layerCount *
          (idx * 2.5 / wMultiple)}px)`
      }))
    })
  }

  handleTouchMove = evt => {
    evt.preventDefault()
    const {pageX, pageY} = evt.touches[0]
    this.handleMove({pageX, pageY})
  }

  handleEnter = () => {
    this.setState({isOnHover: true})
  }

  handleLeave = () => {
    this.setState({
      isOnHover: false,
      container: {},
      shine: {},
      layers: []
    })
  }

  renderShadow = () => (
    <div
      style={{
        ...styles.shadow,
        ...(this.state.isOnHover ? styles.shadowOnHover : {})
      }}
    />
  )

  renderLayers = () => (
    <div style={styles.layers}>
      {this.props.layers &&
        this.props.layers.map((imgSrc, idx) => (
          <div
            style={{
              backgroundImage: `url(${imgSrc})`,
              ...styles.renderedLayer,
              ...(this.state.layers[idx] ? this.state.layers[idx] : {})
            }}
            key={idx}
          />
        ))}
    </div>
  )

  renderShine = () => <div style={{...styles.shine, ...this.state.shine}} />

  render () {
    if (this.props.isStatic) {
      return (
        <div
          style={{
            ...styles.root,
            ...(this.props.style ? this.props.style : {})
          }}
          className={this.props.className || ''}
        >
          <img style={styles.staticFallback} src={this.props.staticFallback} />
        </div>
      )
    }

    return (
      <div
        style={{
          ...styles.root,
          transform: `perspective(${this.state.rootElemWidth * 3}px)`,
          ...(this.props.style ? this.props.style : {})
        }}
        onMouseMove={this.handleMove}
        onMouseEnter={this.handleEnter}
        onMouseLeave={this.handleLeave}
        onTouchMove={this.handleTouchMove}
        onTouchStart={this.handleEnter}
        onTouchEnd={this.handleLeave}
        className={this.props.className || ''}
        ref={node => (this.node = node)}
      >
        <div style={{...styles.container, ...this.state.container}}>
          {this.renderShadow()}
          {this.renderLayers()}
          {this.renderShine()}
        </div>
      </div>
    )
  }
}
